<!DOCTYPE html
PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<html>
<head>
<title>BRESEQ :: Summary Statistics</title>
<style type = "text/css">
body {font-family: sans-serif; font-size: 11pt;}
th {background-color: rgb(0,0,0); color: rgb(255,255,255);}
table {background-color: rgb(1,0,0); color: rgb(0,0,0);}
tr {background-color: rgb(255,255,255);}
.mutation_in_codon {color:red; text-decoration : underline;}
.mutation_header_row {background-color: rgb(0,130,0);}
.read_alignment_header_row {background-color: rgb(255,0,0);}
.missing_coverage_header_row {background-color: rgb(0,100,100);}
.new_junction_header_row {background-color: rgb(0,0,155);}
.copy_number_header_row {background-color: rgb(153,102,0);}
.alternate_table_row_0 {background-color: rgb(255,255,255);}
.alternate_table_row_1 {background-color: rgb(235,235,235);}
.gray_table_row {background-color: rgb(230,230,245);}
.polymorphism_table_row {background-color: rgb(160,255,160);}
.highlight_table_row {background-color: rgb(192,255,255);}
.reject_table_row {background-color: rgb(255,200,165);}
.information_table_row {background-color: rgb(200,255,255);}
.junction_repeat {background-color: rgb(255,165,0)}
.junction_gene {}
.hidden { display: none; }
.unhidden { display: block; }

</style>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<script type="text/javascript">
  function hideTog(divID) {
    var item = document.getElementById(divID);
    if (item) {
      item.className=(item.className=='hidden')?'unhidden':'hidden';
    }
  }
  function showTog(butID) {
    var button = document.getElementById(butID);
    if (button) {
      button.value=(button.value=='Show')?'Hide':'Show';
    }
  }
</script>

</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="3">
<tr>
<td><a href="http://barricklab.org/breseq"><img src="evidence/breseq_small.png" /></a></td>
<td width="100%">
<b><i>breseq</i></b>&nbsp;&nbsp;version 0.24rc1&nbsp;&nbsp;revision 83e50355436e
<br><a href="index.html">mutation predictions</a> | 
<a href="marginal.html">marginal predictions</a> | 
<a href="summary.html">summary statistics</a> | 
<a href="output.gd">genome diff</a> | 
<a href="log.txt">command line log</a>
</td></tr></table>

<p>
<h2>Read File Information</h2>
<table border="0" cellspace="1" cellpadding="5">
<tr ><th></th><th>read file</th><th>reads</th><th>bases</th><th>passed&nbsp;filters</th><th>average</th><th>longest</th><th>mapped</th></tr>
<tr ><td><a href="calibration/SRR030257_1.error_rates.pdf">errors</a></td><td>SRR030257_1</td><td align="right">3,739,210</td><td align="right">134,611,560</td><td align="right">98.4%</td><td align="right">36.0&nbsp;bases</td><td align="right">36&nbsp;bases</td><td align="right">95.1%</td></tr><tr ><td><a href="calibration/SRR030257_2.error_rates.pdf">errors</a></td><td>SRR030257_2</td><td align="right">3,752,181</td><td align="right">135,078,516</td><td align="right">98.7%</td><td align="right">36.0&nbsp;bases</td><td align="right">36&nbsp;bases</td><td align="right">93.9%</td></tr><tr class="highlight_table_row"><td></td><td><b>total</b></td><td align="right"><b>7,491,391</b></td><td align="right"><b>269,690,076</b></td><td align="right">98.6%</td><td align="right">36.0&nbsp;bases</td><td><b>36</b>&nbsp;bases</td><td align="right">94.5%</td></tr></table><h2>Reference Sequence Information</h2>
<p>
<table border="0" cellspacing="1" cellpadding="5" >
<tr><th></th><th></th><th>seq id</th><th>length</th><th align="center">fit mean</th><th align="center">fit dispersion</th><th align="left">description</th></tr>
<tr><td><a href="evidence/REL606.overview.png">coverage</a></td><td><a href="calibration/REL606.unique_coverage.pdf">distribution</a></td><td>REL606</td><td align="right">4,629,812</td><td align="center">55.2</td><td align="center">3.0</td><td>Escherichia coli strain REL606.</td></tr><tr class="highlight_table_row"><td></td><td></td><td><b>total</b></td><td align="right"><b>4,629,812</b></td><td></td><td></td><td></td></tr>
</table>
<p><b>fit dispersion</b> is the ratio of the variance to the mean for the negative binomial fit. It is =1 for Poisson and >1 for over-dispersed data.
<h2>Junction Evidence</h2>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr ><th>reference sequence</th><th>pr(no read start)</th></tr>
<tr >
<td>REL606</td><td align="center">0.51817</td></tr></table><p><b>pr(no read start)</b> is the probability that there will not be an aligned read whose first base matches a given position on a given strand.
<h2>Read Alignment Evidence</h2>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th>option</th><th>value</th></tr><tr><td>Mode</td><td>Consensus/Mixed Base</td></tr><tr><td>Ploidy</td><td>1 (haploid)</td></tr><tr><td>Consensus mutation E-value cutoff</td><td>10</td></tr><tr><td>Polymorphism E-value cutoff</td><td>10</td></tr><tr><td>Polymorphism frequency cutoff</td><td>0.1</td></tr><tr><td>Polymorphism minimum coverage each strand</td><td>2</td></tr><tr><td>Polymorphism bias cutoff</td><td>0.05</td></tr><tr><td>Predict indel polymorphisms</td><td>YES</td></tr><tr><td>Skip indel polymorphisms in homopolymers runs</td><td> &ge;3 bases</td></tr></table><p>
<h2>Execution Times</h2>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th>step</th><th>start</th><th>end</th><th>elapsed</th></tr>
<tr><td>Read and reference sequence file input</td><td>22:43:45&nbsp;01&nbsp;Sep&nbsp;2013</td><td>22:45:37&nbsp;01&nbsp;Sep&nbsp;2013</td><td>1&nbsp;minute&nbsp;52&nbsp;seconds</td></tr>
<tr><td>Read alignment to reference genome</td><td>22:45:37&nbsp;01&nbsp;Sep&nbsp;2013</td><td>02:18:31&nbsp;02&nbsp;Sep&nbsp;2013</td><td>3&nbsp;hours&nbsp;32&nbsp;minutes&nbsp;54&nbsp;seconds</td></tr>
<tr><td>Preprocessing alignments for candidate junction identification</td><td>02:18:31&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:28:28&nbsp;02&nbsp;Sep&nbsp;2013</td><td>9&nbsp;minutes&nbsp;57&nbsp;seconds</td></tr>
<tr><td>Preliminary analysis of coverage distribution</td><td>02:28:28&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:33:26&nbsp;02&nbsp;Sep&nbsp;2013</td><td>4&nbsp;minutes&nbsp;58&nbsp;seconds</td></tr>
<tr><td>Identifying junction candidates</td><td>02:33:26&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:34:02&nbsp;02&nbsp;Sep&nbsp;2013</td><td>36&nbsp;seconds</td></tr>
<tr><td>Re-alignment to junction candidates</td><td>02:34:02&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:36:25&nbsp;02&nbsp;Sep&nbsp;2013</td><td>2&nbsp;minutes&nbsp;23&nbsp;seconds</td></tr>
<tr><td>Resolving alignments with junction candidates</td><td>02:36:26&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:45:11&nbsp;02&nbsp;Sep&nbsp;2013</td><td>8&nbsp;minutes&nbsp;45&nbsp;seconds</td></tr>
<tr><td>Creating BAM files</td><td>02:45:11&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:49:41&nbsp;02&nbsp;Sep&nbsp;2013</td><td>4&nbsp;minutes&nbsp;30&nbsp;seconds</td></tr>
<tr><td>Tabulating error counts</td><td>02:49:41&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:51:10&nbsp;02&nbsp;Sep&nbsp;2013</td><td>1&nbsp;minute&nbsp;29&nbsp;seconds</td></tr>
<tr><td>Re-calibrating base error rates</td><td>02:51:10&nbsp;02&nbsp;Sep&nbsp;2013</td><td>02:51:12&nbsp;02&nbsp;Sep&nbsp;2013</td><td>2&nbsp;seconds</td></tr>
<tr><td>Examining read alignment evidence</td><td>02:51:12&nbsp;02&nbsp;Sep&nbsp;2013</td><td>03:05:55&nbsp;02&nbsp;Sep&nbsp;2013</td><td>14&nbsp;minutes&nbsp;43&nbsp;seconds</td></tr>
<tr><td>Output</td><td>03:05:55&nbsp;02&nbsp;Sep&nbsp;2013</td><td>03:07:32&nbsp;02&nbsp;Sep&nbsp;2013</td><td>1&nbsp;minute&nbsp;37&nbsp;seconds</td></tr>
<tr class="highlight_table_row">
<td colspan="3" ><b>Total</b></td>
<td><b>4&nbsp;hours&nbsp;23&nbsp;minutes&nbsp;46&nbsp;seconds</b></td>
</tr>
</table></body></html>