<!DOCTYPE html
PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<html>
<head>
<title>BRESEQ :: Mutation Predictions</title>
<style type = "text/css">
body {font-family: sans-serif; font-size: 11pt;}
th {background-color: rgb(0,0,0); color: rgb(255,255,255);}
table {background-color: rgb(1,0,0); color: rgb(0,0,0);}
tr {background-color: rgb(255,255,255);}
.mutation_in_codon {color:red; text-decoration : underline;}
.snp_type_synonymous{color:green;}
.snp_type_nonsynonymous{color:blue;}
.snp_type_nonsense{color:red;}
.mutation_header_row {background-color: rgb(0,130,0);}
.read_alignment_header_row {background-color: rgb(255,0,0);}
.missing_coverage_header_row {background-color: rgb(0,100,100);}
.new_junction_header_row {background-color: rgb(0,0,155);}
.copy_number_header_row {background-color: rgb(153,102,0);}
.alternate_table_row_0 {background-color: rgb(255,255,255);}
.alternate_table_row_1 {background-color: rgb(235,235,235);}
.gray_table_row {background-color: rgb(230,230,245);}
.polymorphism_table_row {background-color: rgb(160,255,160);}
.highlight_table_row {background-color: rgb(192,255,255);}
.reject_table_row {background-color: rgb(255,200,165);}
.user_defined_table_row {background-color: rgb(255,255,0);}
.information_table_row {background-color: rgb(200,255,255);}
.junction_repeat {background-color: rgb(255,165,0)}
.junction_gene {}
.hidden { display: none; }
.unhidden { display: block; }

</style>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<script type="text/javascript">
  function hideTog(divID) {
    var item = document.getElementById(divID);
    if (item) {
      item.className=(item.className=='hidden')?'unhidden':'hidden';
    }
  }
  function showTog(butID) {
    var button = document.getElementById(butID);
    if (button) {
      button.value=(button.value=='Show')?'Hide':'Show';
    }
  }
</script>

</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="3">
<tr>
<td><a href="http://barricklab.org/breseq"><img src="evidence/breseq_small.png" /></a></td>
<td width="100%">
<b><i>breseq</i></b>&nbsp;&nbsp;version 0.35.0&nbsp;&nbsp;
<br><a href="index.html">mutation predictions</a> | 
<a href="marginal.html">marginal predictions</a> | 
<a href="summary.html">summary statistics</a> | 
<a href="output.gd">genome diff</a> | 
<a href="log.txt">command line log</a>
</td></tr></table>

<p>
<!--Mutation Predictions -->
<p>
<!--Output Html_Mutation_Table_String-->
<table border="0" cellspacing="1" cellpadding="3">
<tr><th colspan="7" align="left" class="mutation_header_row">Predicted mutations</th></tr><tr>
<th>evidence</th>
<th>seq&nbsp;id</th>
<th>position</th>
<th>mutation</th>
<th>annotation</th>
<th>gene</th>
<th width="100%">description</th>
</tr>

<!-- Item Lines -->

<!-- Print The Table Row -->
<tr class="normal_table_row">
<td align="center"><a href="evidence/SNP_1.html">RA</a></td><!-- Evidence -->
<td align="center">NZ_LR130236</td><!-- Seq_Id -->
<td align="right">222,137</td><!-- Position -->
<td align="center">A&rarr;C</td><!-- Cell Mutation -->
<td align="center"><font class="snp_type_nonsynonymous">E38D</font>&nbsp;(GA<font class="mutation_in_codon">A</font>&rarr;GA<font class="mutation_in_codon">C</font>)&nbsp;</td>
<td align="center"><i>EW021_RS01235</i>&nbsp;&rarr;</td>
<td align="left">alpha&#8209;glucosidase</td>
</tr>
<!-- End Table Row -->

<!-- Print The Table Row -->
<tr class="normal_table_row">
<td align="center"><a href="evidence/SNP_2.html">RA</a></td><!-- Evidence -->
<td align="center">krmit</td><!-- Seq_Id -->
<td align="right">474</td><!-- Position -->
<td align="center">C&rarr;G</td><!-- Cell Mutation -->
<td align="center">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center"><i>&#8211;</i>&nbsp;/&nbsp;<i>&#8211;</i></td>
<td align="left">&#8211;/&#8211;</td>
</tr>
<!-- End Table Row -->

<!-- Print The Table Row -->
<tr class="normal_table_row">
<td align="center"><a href="evidence/DEL_3.html">RA</a></td><!-- Evidence -->
<td align="center">krmit</td><!-- Seq_Id -->
<td align="right">803</td><!-- Position -->
<td align="center">&Delta;1&nbsp;bp</td><!-- Cell Mutation -->
<td align="center">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center"><i>&#8211;</i>&nbsp;/&nbsp;<i>&#8211;</i></td>
<td align="left">&#8211;/&#8211;</td>
</tr>
<!-- End Table Row -->

<!-- Print The Table Row -->
<tr class="normal_table_row">
<td align="center"><a href="evidence/INS_4.html">RA</a></td><!-- Evidence -->
<td align="center">krmit</td><!-- Seq_Id -->
<td align="right">814</td><!-- Position -->
<td align="center">+A</td><!-- Cell Mutation -->
<td align="center">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center"><i>&#8211;</i>&nbsp;/&nbsp;<i>&#8211;</i></td>
<td align="left">&#8211;/&#8211;</td>
</tr>
<!-- End Table Row -->
</table>
<p>
<table border="0" cellspacing="1" cellpadding="3" width="100%">
<tr><th colspan="11" align="left" class="missing_coverage_header_row">Unassigned missing coverage evidence</th></tr>
<tr><th>&nbsp;</th><th>&nbsp;</th><th>&nbsp;</th><th>seq&nbsp;id</th>
<th>start</th>
<th>end</th>
<th>size</th>
<th>&larr;reads</th>
<th>reads&rarr;</th>
<th>gene</th>
<th width="100%">description</th>
</tr>
<tr>
<td><a href="evidence/MC_SIDE_1_9.html">*</a></td>
<td><a href="evidence/MC_SIDE_2_9.html">*</a></td>
<td><a href="evidence/MC_PLOT_9.html">&divide;</a></td>
<td>krmit</td>
<td align="right">1</td>
<td align="right">21</td>
<td align="right">21</td>
<td align="center">NA&nbsp;[0]</td>
<td align="center">[41]&nbsp;235</td>
<td align="center"><i>&#8211;/&#8211;</i></td>
<td align="left">&#8211;/&#8211;</td>
</tr>
<tr>
<td><a href="evidence/MC_SIDE_1_10.html">*</a></td>
<td><a href="evidence/MC_SIDE_2_10.html">*</a></td>
<td><a href="evidence/MC_PLOT_10.html">&divide;</a></td>
<td>krmit</td>
<td align="right">1620</td>
<td align="right">1641</td>
<td align="right">22</td>
<td align="center">297&nbsp;[28]</td>
<td align="center">[0]&nbsp;NA</td>
<td align="center"><i>&#8211;/&#8211;</i></td>
<td align="left">&#8211;/&#8211;</td>
</tr>
</table>
<p>
<table border="0" cellspacing="1" cellpadding="3">
<tr><th colspan="12" align="left" class="new_junction_header_row">Unassigned new junction evidence</th></tr>
<!-- Header Lines for New Junction -->
<tr>
<th colspan="2">&nbsp;</th>
<th>seq&nbsp;id</th>
<th>position</th>
<th>reads&nbsp;(cov)</th>
<th>reads&nbsp;(cov)</th>
<th>score</th>
<th>skew</th>
<th>freq</th>
<th>annotation</th>
<th>gene</th>
<th width="100%">product</th>
</tr>

<!-- Item Lines for New Junction -->
<!-- Side 1 Item Lines for New Junction -->
<tr class="mutation_table_row_0">
<td rowspan="2"><a href="evidence/JC_12.html">*</a></td>
<td rowspan="1"><a href="evidence/JC_SIDE_1_12.html">?</a></td>
<td rowspan="1" class="junction_repeat">NZ_LR130236</td>
<td align="center" class="junction_repeat">16713&nbsp;=</td><td align="center" class="junction_repeat">NA (NA)</td><td rowspan="2" align="center">374 (1.160)<br>+C</td>
<td rowspan="2" align="center">130/252</td>
<td rowspan="2" align="center">0.0</td>
<td rowspan="2" align="center">56.5%</td>
<td align="center" class="junction_repeat">intergenic&nbsp;(+444/&#8209;122)</td>
<td align="center" class="junction_repeat"><i>EW021_RS00065/EW021_RS00070</i></td>
<td class="junction_repeat">amino acid permease/16S ribosomal RNA</td>
</tr>
<!-- Side 2 Item Lines for New Junction -->
<tr class="mutation_table_row_0">
<td rowspan="1"><a href="evidence/JC_SIDE_2_12.html">?</a></td><td rowspan="1" class="junction_gene">NZ_LR130236</td>
<td align="center" class="junction_gene">=&nbsp;22712</td>
<td align="center" class="junction_gene">290 (0.890)</td><td align="center" class="junction_gene">intergenic&nbsp;(+3/&#8209;3)</td>
<td align="center" class="junction_gene"><i>EW021_RS00130/EW021_RS00135</i></td>
<td class="junction_gene">tRNA&#8209;Pro/tRNA&#8209;Met</td>
</tr>
<!-- Side 1 Item Lines for New Junction -->
<tr class="mutation_table_row_1">
<td rowspan="2"><a href="evidence/JC_18.html">*</a></td>
<td rowspan="1"><a href="evidence/JC_SIDE_1_18.html">?</a></td>
<td rowspan="1" class="junction_gene">NZ_LR130236</td>
<td align="center" class="junction_gene">1382207&nbsp;=</td><td align="center" class="junction_gene">181 (0.560)</td><td rowspan="2" align="center">188 (0.510)</td>
<td rowspan="2" align="center">89/246</td>
<td rowspan="2" align="center">0.6</td>
<td rowspan="2" align="center">67.3%</td>
<td align="center" class="junction_gene">coding&nbsp;(451/723&nbsp;nt)</td>
<td align="center" class="junction_gene"><i>EW021_RS07155</i></td>
<td class="junction_gene">hypothetical protein</td>
</tr>
<!-- Side 2 Item Lines for New Junction -->
<tr class="mutation_table_row_1">
<td rowspan="1"><a href="evidence/JC_SIDE_2_18.html">?</a></td><td rowspan="1" class="junction_gene">krmit</td>
<td align="center" class="junction_gene">22&nbsp;=</td>
<td align="center" class="junction_gene">7 (0.020)</td><td align="center" class="junction_gene">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center" class="junction_gene"><i>&#8211;/&#8211;</i></td>
<td class="junction_gene">&#8211;/&#8211;</td>
</tr>
<!-- Side 1 Item Lines for New Junction -->
<tr class="mutation_table_row_0">
<td rowspan="2"><a href="evidence/JC_19.html">*</a></td>
<td rowspan="1"><a href="evidence/JC_SIDE_1_19.html">?</a></td>
<td rowspan="1" class="junction_gene">NZ_LR130236</td>
<td align="center" class="junction_gene">=&nbsp;1382209</td><td align="center" class="junction_gene">180 (0.550)</td><td rowspan="2" align="center">159 (0.430)</td>
<td rowspan="2" align="center">76/250</td>
<td rowspan="2" align="center">1.2</td>
<td rowspan="2" align="center">63.2%</td>
<td align="center" class="junction_gene">coding&nbsp;(453/723&nbsp;nt)</td>
<td align="center" class="junction_gene"><i>EW021_RS07155</i></td>
<td class="junction_gene">hypothetical protein</td>
</tr>
<!-- Side 2 Item Lines for New Junction -->
<tr class="mutation_table_row_0">
<td rowspan="1"><a href="evidence/JC_SIDE_2_19.html">?</a></td><td rowspan="1" class="junction_gene">krmit</td>
<td align="center" class="junction_gene">=&nbsp;1619</td>
<td align="center" class="junction_gene">8 (0.020)</td><td align="center" class="junction_gene">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center" class="junction_gene"><i>&#8211;/&#8211;</i></td>
<td class="junction_gene">&#8211;/&#8211;</td>
</tr>
<!-- Side 1 Item Lines for New Junction -->
<tr class="mutation_table_row_1">
<td rowspan="2"><a href="evidence/JC_30.html">*</a></td>
<td rowspan="1"><a href="evidence/JC_SIDE_1_30.html">?</a></td>
<td rowspan="1" class="junction_gene">NZ_LR130236</td>
<td align="center" class="junction_gene">1807263&nbsp;=</td><td align="center" class="junction_gene">475 (1.460)</td><td rowspan="2" align="center">105 (0.280)</td>
<td rowspan="2" align="center">61/250</td>
<td rowspan="2" align="center">2.0</td>
<td rowspan="2" align="center">30.6%</td>
<td align="center" class="junction_gene">coding&nbsp;(822/1290&nbsp;nt)</td>
<td align="center" class="junction_gene"><i>EW021_RS09285</i></td>
<td class="junction_gene">insulinase family protein</td>
</tr>
<!-- Side 2 Item Lines for New Junction -->
<tr class="mutation_table_row_1">
<td rowspan="1"><a href="evidence/JC_SIDE_2_30.html">?</a></td><td rowspan="1" class="junction_gene">krmit</td>
<td align="center" class="junction_gene">=&nbsp;1619</td>
<td align="center" class="junction_gene">8 (0.020)</td><td align="center" class="junction_gene">intergenic&nbsp;(&#8211;/&#8211;)</td>
<td align="center" class="junction_gene"><i>&#8211;/&#8211;</i></td>
<td class="junction_gene">&#8211;/&#8211;</td>
</tr>
</table>
</body></html>