<!DOCTYPE html
PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<html>
<head>
<title>BRESEQ :: Summary Statistics</title>
<style type = "text/css">
body {font-family: sans-serif; font-size: 11pt;}
th {background-color: rgb(0,0,0); color: rgb(255,255,255);}
table {background-color: rgb(1,0,0); color: rgb(0,0,0);}
tr {background-color: rgb(255,255,255);}
.mutation_in_codon {color:red; text-decoration : underline;}
.snp_type_synonymous{color:green;}
.snp_type_nonsynonymous{color:blue;}
.snp_type_nonsense{color:red;}
.mutation_header_row {background-color: rgb(0,130,0);}
.read_alignment_header_row {background-color: rgb(255,0,0);}
.missing_coverage_header_row {background-color: rgb(0,100,100);}
.new_junction_header_row {background-color: rgb(0,0,155);}
.copy_number_header_row {background-color: rgb(153,102,0);}
.alternate_table_row_0 {background-color: rgb(255,255,255);}
.alternate_table_row_1 {background-color: rgb(235,235,235);}
.gray_table_row {background-color: rgb(230,230,245);}
.polymorphism_table_row {background-color: rgb(160,255,160);}
.highlight_table_row {background-color: rgb(192,255,255);}
.reject_table_row {background-color: rgb(255,200,165);}
.user_defined_table_row {background-color: rgb(255,255,0);}
.information_table_row {background-color: rgb(200,255,255);}
.junction_repeat {background-color: rgb(255,165,0)}
.junction_gene {}
.hidden { display: none; }
.unhidden { display: block; }

</style>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<script type="text/javascript">
  function hideTog(divID) {
    var item = document.getElementById(divID);
    if (item) {
      item.className=(item.className=='hidden')?'unhidden':'hidden';
    }
  }
  function showTog(butID) {
    var button = document.getElementById(butID);
    if (button) {
      button.value=(button.value=='Show')?'Hide':'Show';
    }
  }
</script>

</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="3">
<tr>
<td><a href="http://barricklab.org/breseq"><img src="evidence/breseq_small.png" /></a></td>
<td width="100%">
<b><i>breseq</i></b>&nbsp;&nbsp;version 0.32.1&nbsp;&nbsp;
<br><a href="index.html">mutation predictions</a> | 
<a href="marginal.html">marginal predictions</a> | 
<a href="summary.html">summary statistics</a> | 
<a href="output.gd">genome diff</a> | 
<a href="log.txt">command line log</a>
</td></tr></table>

<p>
<h2>Read File Information</h2>
<table border="0" cellspace="1" cellpadding="5">
<tr ><th></th><th>read file</th><th>reads</th><th>bases</th><th>passed&nbsp;filters</th><th>average</th><th>longest</th><th>mapped</th></tr>
<tr ><td><a href="calibration/SC1210.forward.trimmed.paired.error_rates.pdf">errors</a></td><td>SC1210.forward.trimmed.paired</td><td align="right">3,740,672</td><td align="right">483,102,443</td><td align="right">100.0%</td><td align="right">129.1&nbsp;bases</td><td align="right">151&nbsp;bases</td><td align="right">29.9%</td></tr><tr ><td><a href="calibration/SC1210.reverse.trimmed.paired.error_rates.pdf">errors</a></td><td>SC1210.reverse.trimmed.paired</td><td align="right">3,739,319</td><td align="right">487,767,242</td><td align="right">100.0%</td><td align="right">130.4&nbsp;bases</td><td align="right">151&nbsp;bases</td><td align="right">29.7%</td></tr><tr ><td><a href="calibration/SC1210.forward.trimmed.unpaired.error_rates.pdf">errors</a></td><td>SC1210.forward.trimmed.unpaired</td><td align="right">79,254</td><td align="right">8,686,784</td><td align="right">100.0%</td><td align="right">109.6&nbsp;bases</td><td align="right">151&nbsp;bases</td><td align="right">33.1%</td></tr><tr class="highlight_table_row"><td></td><td><b>total</b></td><td align="right"><b>7,559,245</b></td><td align="right"><b>979,556,469</b></td><td align="right">100.0%</td><td align="right">129.6&nbsp;bases</td><td><b>151</b>&nbsp;bases</td><td align="right">29.9%</td></tr></table><h2>Reference Sequence Information</h2>
<p>
<table border="0" cellspacing="1" cellpadding="5" >
<tr><th></th><th></th><th>seq id</th><th>length</th><th align="center">fit mean</th><th align="center">fit dispersion</th><th align="center">% mapped reads</th><th align="left">description</th></tr>
<tr><td><a href="evidence/contig00001.overview.png">coverage</a></td><td><a href="calibration/0.unique_coverage.pdf">distribution</a></td><td>contig00001</td><td align="right">560,358</td><td align="center">38.9</td><td align="center">4.9</td><td align="center">7.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00002.overview.png">coverage</a></td><td><a href="calibration/1.unique_coverage.pdf">distribution</a></td><td>contig00002</td><td align="right">367,866</td><td align="center">35.6</td><td align="center">4.7</td><td align="center">4.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00003.overview.png">coverage</a></td><td><a href="calibration/2.unique_coverage.pdf">distribution</a></td><td>contig00003</td><td align="right">356,113</td><td align="center">38.3</td><td align="center">4.7</td><td align="center">4.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00004.overview.png">coverage</a></td><td><a href="calibration/3.unique_coverage.pdf">distribution</a></td><td>contig00004</td><td align="right">283,025</td><td align="center">38.8</td><td align="center">5.8</td><td align="center">3.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00005.overview.png">coverage</a></td><td><a href="calibration/4.unique_coverage.pdf">distribution</a></td><td>contig00005</td><td align="right">252,402</td><td align="center">38.2</td><td align="center">4.4</td><td align="center">3.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00006.overview.png">coverage</a></td><td><a href="calibration/5.unique_coverage.pdf">distribution</a></td><td>contig00006</td><td align="right">243,346</td><td align="center">39.7</td><td align="center">4.0</td><td align="center">3.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00007.overview.png">coverage</a></td><td><a href="calibration/6.unique_coverage.pdf">distribution</a></td><td>contig00007</td><td align="right">238,053</td><td align="center">41.8</td><td align="center">5.0</td><td align="center">3.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00008.overview.png">coverage</a></td><td><a href="calibration/7.unique_coverage.pdf">distribution</a></td><td>contig00008</td><td align="right">236,843</td><td align="center">40.6</td><td align="center">5.4</td><td align="center">3.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00009.overview.png">coverage</a></td><td><a href="calibration/8.unique_coverage.pdf">distribution</a></td><td>contig00009</td><td align="right">223,110</td><td align="center">38.4</td><td align="center">5.9</td><td align="center">2.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00010.overview.png">coverage</a></td><td><a href="calibration/9.unique_coverage.pdf">distribution</a></td><td>contig00010</td><td align="right">220,017</td><td align="center">34.6</td><td align="center">4.7</td><td align="center">2.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00011.overview.png">coverage</a></td><td><a href="calibration/10.unique_coverage.pdf">distribution</a></td><td>contig00011</td><td align="right">215,188</td><td align="center">42.9</td><td align="center">5.8</td><td align="center">3.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00012.overview.png">coverage</a></td><td><a href="calibration/11.unique_coverage.pdf">distribution</a></td><td>contig00012</td><td align="right">166,255</td><td align="center">36.2</td><td align="center">5.2</td><td align="center">2.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00013.overview.png">coverage</a></td><td><a href="calibration/12.unique_coverage.pdf">distribution</a></td><td>contig00013</td><td align="right">160,860</td><td align="center">46.8</td><td align="center">5.5</td><td align="center">2.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00014.overview.png">coverage</a></td><td><a href="calibration/13.unique_coverage.pdf">distribution</a></td><td>contig00014</td><td align="right">160,619</td><td align="center">48.5</td><td align="center">6.5</td><td align="center">2.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00015.overview.png">coverage</a></td><td><a href="calibration/14.unique_coverage.pdf">distribution</a></td><td>contig00015</td><td align="right">147,609</td><td align="center">36.5</td><td align="center">4.7</td><td align="center">1.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00016.overview.png">coverage</a></td><td><a href="calibration/15.unique_coverage.pdf">distribution</a></td><td>contig00016</td><td align="right">145,556</td><td align="center">35.8</td><td align="center">5.2</td><td align="center">1.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00017.overview.png">coverage</a></td><td><a href="calibration/16.unique_coverage.pdf">distribution</a></td><td>contig00017</td><td align="right">129,672</td><td align="center">37.3</td><td align="center">5.9</td><td align="center">1.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00018.overview.png">coverage</a></td><td><a href="calibration/17.unique_coverage.pdf">distribution</a></td><td>contig00018</td><td align="right">127,784</td><td align="center">42.1</td><td align="center">5.5</td><td align="center">1.8%</td><td>na</td></tr><tr><td><a href="evidence/contig00019.overview.png">coverage</a></td><td><a href="calibration/18.unique_coverage.pdf">distribution</a></td><td>contig00019</td><td align="right">127,184</td><td align="center">40.6</td><td align="center">4.4</td><td align="center">1.8%</td><td>na</td></tr><tr><td><a href="evidence/contig00020.overview.png">coverage</a></td><td><a href="calibration/19.unique_coverage.pdf">distribution</a></td><td>contig00020</td><td align="right">127,133</td><td align="center">38.2</td><td align="center">5.0</td><td align="center">1.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00021.overview.png">coverage</a></td><td><a href="calibration/20.unique_coverage.pdf">distribution</a></td><td>contig00021</td><td align="right">125,593</td><td align="center">38.4</td><td align="center">4.5</td><td align="center">1.8%</td><td>na</td></tr><tr><td><a href="evidence/contig00022.overview.png">coverage</a></td><td><a href="calibration/21.unique_coverage.pdf">distribution</a></td><td>contig00022</td><td align="right">116,943</td><td align="center">37.6</td><td align="center">4.5</td><td align="center">1.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00023.overview.png">coverage</a></td><td><a href="calibration/22.unique_coverage.pdf">distribution</a></td><td>contig00023</td><td align="right">116,289</td><td align="center">43.4</td><td align="center">6.6</td><td align="center">1.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00024.overview.png">coverage</a></td><td><a href="calibration/23.unique_coverage.pdf">distribution</a></td><td>contig00024</td><td align="right">115,038</td><td align="center">38.6</td><td align="center">4.7</td><td align="center">1.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00025.overview.png">coverage</a></td><td><a href="calibration/24.unique_coverage.pdf">distribution</a></td><td>contig00025</td><td align="right">113,691</td><td align="center">46.7</td><td align="center">7.2</td><td align="center">2.2%</td><td>na</td></tr><tr><td><a href="evidence/contig00026.overview.png">coverage</a></td><td><a href="calibration/25.unique_coverage.pdf">distribution</a></td><td>contig00026</td><td align="right">113,040</td><td align="center">44.7</td><td align="center">4.6</td><td align="center">1.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00027.overview.png">coverage</a></td><td><a href="calibration/26.unique_coverage.pdf">distribution</a></td><td>contig00027</td><td align="right">107,209</td><td align="center">34.0</td><td align="center">3.6</td><td align="center">1.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00028.overview.png">coverage</a></td><td><a href="calibration/27.unique_coverage.pdf">distribution</a></td><td>contig00028</td><td align="right">98,304</td><td align="center">35.5</td><td align="center">4.7</td><td align="center">1.2%</td><td>na</td></tr><tr><td><a href="evidence/contig00029.overview.png">coverage</a></td><td><a href="calibration/28.unique_coverage.pdf">distribution</a></td><td>contig00029</td><td align="right">90,842</td><td align="center">61.7</td><td align="center">6.0</td><td align="center">1.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00030.overview.png">coverage</a></td><td><a href="calibration/29.unique_coverage.pdf">distribution</a></td><td>contig00030</td><td align="right">86,056</td><td align="center">35.7</td><td align="center">4.1</td><td align="center">1.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00031.overview.png">coverage</a></td><td><a href="calibration/30.unique_coverage.pdf">distribution</a></td><td>contig00031</td><td align="right">84,995</td><td align="center">40.1</td><td align="center">4.2</td><td align="center">1.2%</td><td>na</td></tr><tr><td><a href="evidence/contig00032.overview.png">coverage</a></td><td><a href="calibration/31.unique_coverage.pdf">distribution</a></td><td>contig00032</td><td align="right">82,144</td><td align="center">51.1</td><td align="center">9.6</td><td align="center">1.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00033.overview.png">coverage</a></td><td><a href="calibration/32.unique_coverage.pdf">distribution</a></td><td>contig00033</td><td align="right">81,248</td><td align="center">31.1</td><td align="center">4.6</td><td align="center">0.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00034.overview.png">coverage</a></td><td><a href="calibration/33.unique_coverage.pdf">distribution</a></td><td>contig00034</td><td align="right">78,840</td><td align="center">33.5</td><td align="center">6.1</td><td align="center">0.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00035.overview.png">coverage</a></td><td><a href="calibration/34.unique_coverage.pdf">distribution</a></td><td>contig00035</td><td align="right">78,574</td><td align="center">41.4</td><td align="center">3.6</td><td align="center">1.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00036.overview.png">coverage</a></td><td><a href="calibration/35.unique_coverage.pdf">distribution</a></td><td>contig00036</td><td align="right">73,212</td><td align="center">42.3</td><td align="center">5.5</td><td align="center">1.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00037.overview.png">coverage</a></td><td><a href="calibration/36.unique_coverage.pdf">distribution</a></td><td>contig00037</td><td align="right">72,493</td><td align="center">35.2</td><td align="center">3.9</td><td align="center">0.9%</td><td>na</td></tr><tr><td><a href="evidence/contig00038.overview.png">coverage</a></td><td><a href="calibration/37.unique_coverage.pdf">distribution</a></td><td>contig00038</td><td align="right">67,924</td><td align="center">40.3</td><td align="center">4.6</td><td align="center">1.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00039.overview.png">coverage</a></td><td><a href="calibration/38.unique_coverage.pdf">distribution</a></td><td>contig00039</td><td align="right">67,439</td><td align="center">35.3</td><td align="center">5.0</td><td align="center">0.8%</td><td>na</td></tr><tr><td><a href="evidence/contig00040.overview.png">coverage</a></td><td><a href="calibration/39.unique_coverage.pdf">distribution</a></td><td>contig00040</td><td align="right">66,045</td><td align="center">42.9</td><td align="center">5.5</td><td align="center">1.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00041.overview.png">coverage</a></td><td><a href="calibration/40.unique_coverage.pdf">distribution</a></td><td>contig00041</td><td align="right">63,374</td><td align="center">35.1</td><td align="center">5.0</td><td align="center">0.8%</td><td>na</td></tr><tr><td><a href="evidence/contig00042.overview.png">coverage</a></td><td><a href="calibration/41.unique_coverage.pdf">distribution</a></td><td>contig00042</td><td align="right">60,590</td><td align="center">56.1</td><td align="center">8.3</td><td align="center">1.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00043.overview.png">coverage</a></td><td><a href="calibration/42.unique_coverage.pdf">distribution</a></td><td>contig00043</td><td align="right">46,847</td><td align="center">37.0</td><td align="center">3.6</td><td align="center">0.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00044.overview.png">coverage</a></td><td><a href="calibration/43.unique_coverage.pdf">distribution</a></td><td>contig00044</td><td align="right">46,739</td><td align="center">43.8</td><td align="center">5.2</td><td align="center">0.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00045.overview.png">coverage</a></td><td><a href="calibration/44.unique_coverage.pdf">distribution</a></td><td>contig00045</td><td align="right">46,308</td><td align="center">44.3</td><td align="center">6.8</td><td align="center">0.7%</td><td>na</td></tr><tr><td><a href="evidence/contig00046.overview.png">coverage</a></td><td><a href="calibration/45.unique_coverage.pdf">distribution</a></td><td>contig00046</td><td align="right">45,584</td><td align="center">31.1</td><td align="center">4.6</td><td align="center">0.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00047.overview.png">coverage</a></td><td><a href="calibration/46.unique_coverage.pdf">distribution</a></td><td>contig00047</td><td align="right">45,237</td><td align="center">38.9</td><td align="center">5.6</td><td align="center">0.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00048.overview.png">coverage</a></td><td><a href="calibration/47.unique_coverage.pdf">distribution</a></td><td>contig00048</td><td align="right">45,086</td><td align="center">33.7</td><td align="center">4.9</td><td align="center">0.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00049.overview.png">coverage</a></td><td><a href="calibration/48.unique_coverage.pdf">distribution</a></td><td>contig00049</td><td align="right">43,937</td><td align="center">37.2</td><td align="center">6.7</td><td align="center">0.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00050.overview.png">coverage</a></td><td><a href="calibration/49.unique_coverage.pdf">distribution</a></td><td>contig00050</td><td align="right">42,128</td><td align="center">39.8</td><td align="center">3.8</td><td align="center">0.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00051.overview.png">coverage</a></td><td><a href="calibration/50.unique_coverage.pdf">distribution</a></td><td>contig00051</td><td align="right">40,690</td><td align="center">39.8</td><td align="center">3.3</td><td align="center">0.6%</td><td>na</td></tr><tr><td><a href="evidence/contig00052.overview.png">coverage</a></td><td><a href="calibration/51.unique_coverage.pdf">distribution</a></td><td>contig00052</td><td align="right">39,193</td><td align="center">38.9</td><td align="center">4.0</td><td align="center">0.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00053.overview.png">coverage</a></td><td><a href="calibration/52.unique_coverage.pdf">distribution</a></td><td>contig00053</td><td align="right">37,769</td><td align="center">35.7</td><td align="center">3.8</td><td align="center">0.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00054.overview.png">coverage</a></td><td><a href="calibration/53.unique_coverage.pdf">distribution</a></td><td>contig00054</td><td align="right">36,662</td><td align="center">34.3</td><td align="center">3.9</td><td align="center">0.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00055.overview.png">coverage</a></td><td><a href="calibration/54.unique_coverage.pdf">distribution</a></td><td>contig00055</td><td align="right">35,498</td><td align="center">41.6</td><td align="center">5.4</td><td align="center">0.5%</td><td>na</td></tr><tr><td><a href="evidence/contig00056.overview.png">coverage</a></td><td><a href="calibration/55.unique_coverage.pdf">distribution</a></td><td>contig00056</td><td align="right">30,180</td><td align="center">41.2</td><td align="center">2.9</td><td align="center">0.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00057.overview.png">coverage</a></td><td><a href="calibration/56.unique_coverage.pdf">distribution</a></td><td>contig00057</td><td align="right">25,308</td><td align="center">35.0</td><td align="center">2.9</td><td align="center">0.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00058.overview.png">coverage</a></td><td><a href="calibration/57.unique_coverage.pdf">distribution</a></td><td>contig00058</td><td align="right">22,471</td><td align="center">36.0</td><td align="center">3.7</td><td align="center">0.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00059.overview.png">coverage</a></td><td><a href="calibration/58.unique_coverage.pdf">distribution</a></td><td>contig00059</td><td align="right">21,956</td><td align="center">35.0</td><td align="center">3.1</td><td align="center">0.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00060.overview.png">coverage</a></td><td><a href="calibration/59.unique_coverage.pdf">distribution</a></td><td>contig00060</td><td align="right">19,222</td><td align="center">41.0</td><td align="center">5.3</td><td align="center">0.3%</td><td>na</td></tr><tr><td><a href="evidence/contig00061.overview.png">coverage</a></td><td><a href="calibration/60.unique_coverage.pdf">distribution</a></td><td>contig00061</td><td align="right">14,814</td><td align="center">71.7</td><td align="center">5.4</td><td align="center">0.4%</td><td>na</td></tr><tr><td><a href="evidence/contig00062.overview.png">coverage</a></td><td><a href="calibration/61.unique_coverage.pdf">distribution</a></td><td>contig00062</td><td align="right">11,318</td><td align="center">41.8</td><td align="center">4.4</td><td align="center">0.2%</td><td>na</td></tr><tr><td><a href="evidence/contig00063.overview.png">coverage</a></td><td><a href="calibration/62.unique_coverage.pdf">distribution</a></td><td>contig00063</td><td align="right">8,141</td><td align="center">29.7</td><td align="center">4.3</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00064.overview.png">coverage</a></td><td><a href="calibration/63.unique_coverage.pdf">distribution</a></td><td>contig00064</td><td align="right">7,460</td><td align="center">42.3</td><td align="center">4.7</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00065.overview.png">coverage</a></td><td><a href="calibration/64.unique_coverage.pdf">distribution</a></td><td>contig00065</td><td align="right">5,645</td><td align="center">34.5</td><td align="center">7.7</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00066.overview.png">coverage</a></td><td><a href="calibration/65.unique_coverage.pdf">distribution</a></td><td>contig00066</td><td align="right">5,317</td><td align="center">594.8</td><td align="center">45.4</td><td align="center">1.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00067.overview.png">coverage</a></td><td><a href="calibration/66.unique_coverage.pdf">distribution</a></td><td>contig00067</td><td align="right">4,625</td><td align="center">53.8</td><td align="center">6.2</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00068.overview.png">coverage</a></td><td><a href="calibration/67.unique_coverage.pdf">distribution</a></td><td>contig00068</td><td align="right">4,246</td><td align="center">40.7</td><td align="center">3.2</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00069.overview.png">coverage</a></td><td><a href="calibration/68.unique_coverage.pdf">distribution</a></td><td>contig00069</td><td align="right">3,888</td><td align="center">56.7</td><td align="center">3.5</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00070.overview.png">coverage</a></td><td><a href="calibration/69.unique_coverage.pdf">distribution</a></td><td>contig00070</td><td align="right">3,117</td><td align="center">27.4</td><td align="center">10.2</td><td align="center">0.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00071.overview.png">coverage</a></td><td><a href="calibration/70.unique_coverage.pdf">distribution</a></td><td>contig00071</td><td align="right">2,392</td><td align="center">194.4</td><td align="center">6.0</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00072.overview.png">coverage</a></td><td><a href="calibration/71.unique_coverage.pdf">distribution</a></td><td>contig00072</td><td align="right">2,312</td><td align="center">69.1</td><td align="center">4.0</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00073.overview.png">coverage</a></td><td><a href="calibration/72.unique_coverage.pdf">distribution</a></td><td>contig00073</td><td align="right">1,863</td><td align="center">159.6</td><td align="center">1.6</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00074.overview.png">coverage</a></td><td><a href="calibration/73.unique_coverage.pdf">distribution</a></td><td>contig00074</td><td align="right">1,457</td><td align="center">190.5</td><td align="center">2.7</td><td align="center">0.1%</td><td>na</td></tr><tr class="reject_table_row"><td><a href="evidence/contig00075.overview.png">coverage</a></td><td><a href="calibration/74.unique_coverage.pdf">distribution</a></td><td>contig00075</td><td align="right">1,355</td><td align="center">[268.5]</td><td align="center">[12.6]</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00076.overview.png">coverage</a></td><td><a href="calibration/75.unique_coverage.pdf">distribution</a></td><td>contig00076</td><td align="right">1,355</td><td align="center">178.7</td><td align="center">6.1</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00077.overview.png">coverage</a></td><td><a href="calibration/76.unique_coverage.pdf">distribution</a></td><td>contig00077</td><td align="right">1,345</td><td align="center">97.1</td><td align="center">2.5</td><td align="center">0.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00078.overview.png">coverage</a></td><td><a href="calibration/77.unique_coverage.pdf">distribution</a></td><td>contig00078</td><td align="right">1,289</td><td align="center">59.9</td><td align="center">4.4</td><td align="center">0.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00079.overview.png">coverage</a></td><td><a href="calibration/78.unique_coverage.pdf">distribution</a></td><td>contig00079</td><td align="right">1,229</td><td align="center">255.1</td><td align="center">3.3</td><td align="center">0.1%</td><td>na</td></tr><tr><td><a href="evidence/contig00080.overview.png">coverage</a></td><td><a href="calibration/79.unique_coverage.pdf">distribution</a></td><td>contig00080</td><td align="right">1,195</td><td align="center">102.9</td><td align="center">7.3</td><td align="center">0.0%</td><td>na</td></tr><tr><td><a href="evidence/contig00081.overview.png">coverage</a></td><td><a href="calibration/80.unique_coverage.pdf">distribution</a></td><td>contig00081</td><td align="right">1,189</td><td align="center">85.0</td><td align="center">1.6</td><td align="center">0.0%</td><td>na</td></tr><tr class="highlight_table_row"><td></td><td></td><td><b>total</b></td><td align="right"><b>7,205,243</b></td><td></td><td></td><td align="center">100.0%</td><td></td></tr>
</table>
<p><b>fit dispersion</b> is the ratio of the variance to the mean for the negative binomial fit. It is =1 for Poisson and >1 for over-dispersed data.
<p><span class="reject_table_row">Fit failed</span> Negative binomial fit failed for this reference sequence. It may have an unusual coverage depth distribution. JC and MC predictions may be less accurate.
<h2>New Junction Evidence</h2>
<p>
<h3>Junction Candidates Tested</h3>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th align="center">option</th><th align="center">limit</th><th align="center">actual</th></tr><tr><td>Number of alignment pairs examined for constructing junction candidates</td><td align="center">&le;&nbsp;100000</td><td align="center">1637</td></tr><tr><td>Coverage evenness (position-hash) score of junction candidates</td><td align="center">&ge;&nbsp;2</td><td align="center">&ge;&nbsp;2</td></tr><tr><td>Test this many junction candidates (n). May be smaller if not enough passed the coverage evenness threshold</td><td align="center">100&nbsp;&le;&nbsp;n&nbsp;&le;&nbsp;5000</td><td align="center">227</td></tr><tr><td>Total length of all junction candidates (factor times the reference genome length)</td><td align="center">&le;&nbsp;0.1</td><td align="center">0.010</td></tr></table>
<h3>Junction Skew Score Calculation</h3>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr ><th>reference sequence</th><th>pr(no read start)</th></tr>
<tr >
<td>contig00001</td><td align="center">0.88564</td></tr><tr >
<td>contig00002</td><td align="center">0.89685</td></tr><tr >
<td>contig00003</td><td align="center">0.88727</td></tr><tr >
<td>contig00004</td><td align="center">0.88782</td></tr><tr >
<td>contig00005</td><td align="center">0.88714</td></tr><tr >
<td>contig00006</td><td align="center">0.88519</td></tr><tr >
<td>contig00007</td><td align="center">0.88127</td></tr><tr >
<td>contig00008</td><td align="center">0.88173</td></tr><tr >
<td>contig00009</td><td align="center">0.88895</td></tr><tr >
<td>contig00010</td><td align="center">0.89738</td></tr><tr >
<td>contig00011</td><td align="center">0.87927</td></tr><tr >
<td>contig00012</td><td align="center">0.89835</td></tr><tr >
<td>contig00013</td><td align="center">0.86672</td></tr><tr >
<td>contig00014</td><td align="center">0.86206</td></tr><tr >
<td>contig00015</td><td align="center">0.89137</td></tr><tr >
<td>contig00016</td><td align="center">0.89081</td></tr><tr >
<td>contig00017</td><td align="center">0.89538</td></tr><tr >
<td>contig00018</td><td align="center">0.87967</td></tr><tr >
<td>contig00019</td><td align="center">0.88207</td></tr><tr >
<td>contig00020</td><td align="center">0.88816</td></tr><tr >
<td>contig00021</td><td align="center">0.88202</td></tr><tr >
<td>contig00022</td><td align="center">0.88914</td></tr><tr >
<td>contig00023</td><td align="center">0.87928</td></tr><tr >
<td>contig00024</td><td align="center">0.88782</td></tr><tr >
<td>contig00025</td><td align="center">0.85219</td></tr><tr >
<td>contig00026</td><td align="center">0.86453</td></tr><tr >
<td>contig00027</td><td align="center">0.89810</td></tr><tr >
<td>contig00028</td><td align="center">0.89715</td></tr><tr >
<td>contig00029</td><td align="center">0.84020</td></tr><tr >
<td>contig00030</td><td align="center">0.89655</td></tr><tr >
<td>contig00031</td><td align="center">0.88492</td></tr><tr >
<td>contig00032</td><td align="center">0.86490</td></tr><tr >
<td>contig00033</td><td align="center">0.90490</td></tr><tr >
<td>contig00034</td><td align="center">0.90327</td></tr><tr >
<td>contig00035</td><td align="center">0.87802</td></tr><tr >
<td>contig00036</td><td align="center">0.88007</td></tr><tr >
<td>contig00037</td><td align="center">0.89401</td></tr><tr >
<td>contig00038</td><td align="center">0.88114</td></tr><tr >
<td>contig00039</td><td align="center">0.89907</td></tr><tr >
<td>contig00040</td><td align="center">0.87178</td></tr><tr >
<td>contig00041</td><td align="center">0.88975</td></tr><tr >
<td>contig00042</td><td align="center">0.84913</td></tr><tr >
<td>contig00043</td><td align="center">0.88301</td></tr><tr >
<td>contig00044</td><td align="center">0.87005</td></tr><tr >
<td>contig00045</td><td align="center">0.87761</td></tr><tr >
<td>contig00046</td><td align="center">0.90189</td></tr><tr >
<td>contig00047</td><td align="center">0.88891</td></tr><tr >
<td>contig00048</td><td align="center">0.90370</td></tr><tr >
<td>contig00049</td><td align="center">0.88956</td></tr><tr >
<td>contig00050</td><td align="center">0.88439</td></tr><tr >
<td>contig00051</td><td align="center">0.88246</td></tr><tr >
<td>contig00052</td><td align="center">0.88607</td></tr><tr >
<td>contig00053</td><td align="center">0.88983</td></tr><tr >
<td>contig00054</td><td align="center">0.89882</td></tr><tr >
<td>contig00055</td><td align="center">0.88461</td></tr><tr >
<td>contig00056</td><td align="center">0.87693</td></tr><tr >
<td>contig00057</td><td align="center">0.88934</td></tr><tr >
<td>contig00058</td><td align="center">0.89178</td></tr><tr >
<td>contig00059</td><td align="center">0.89004</td></tr><tr >
<td>contig00060</td><td align="center">0.88064</td></tr><tr >
<td>contig00061</td><td align="center">0.82408</td></tr><tr >
<td>contig00062</td><td align="center">0.87498</td></tr><tr >
<td>contig00063</td><td align="center">0.89641</td></tr><tr >
<td>contig00064</td><td align="center">0.87799</td></tr><tr >
<td>contig00065</td><td align="center">0.90328</td></tr><tr >
<td>contig00066</td><td align="center">0.42180</td></tr><tr >
<td>contig00067</td><td align="center">0.86276</td></tr><tr >
<td>contig00068</td><td align="center">0.89368</td></tr><tr >
<td>contig00069</td><td align="center">0.84820</td></tr><tr >
<td>contig00070</td><td align="center">0.93746</td></tr><tr >
<td>contig00071</td><td align="center">0.63555</td></tr><tr >
<td>contig00072</td><td align="center">0.82745</td></tr><tr >
<td>contig00073</td><td align="center">0.68565</td></tr><tr >
<td>contig00074</td><td align="center">0.64014</td></tr><tr >
<td>contig00075</td><td align="center">NA</td></tr><tr >
<td>contig00076</td><td align="center">0.65166</td></tr><tr >
<td>contig00077</td><td align="center">0.79205</td></tr><tr >
<td>contig00078</td><td align="center">0.84085</td></tr><tr >
<td>contig00079</td><td align="center">0.59277</td></tr><tr >
<td>contig00080</td><td align="center">0.78053</td></tr><tr >
<td>contig00081</td><td align="center">0.81142</td></tr></table><p><b>pr(no read start)</b> is the probability that there will not be an aligned read whose first base matches a given position on a given strand.
<h3>Final Junction Predictions</h3>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th align="center">option</th><th align="center">value</th></tr><tr><td>Coverage evenness (position-hash) score of predicted junctions must be</td><td align="center">&ge;&nbsp;3</td></tr><tr><td>Skew score of predicted junction (&minus;log10 probability of unusual coverage evenness) must be</td><td align="center">&le;&nbsp;3</td></tr><tr><td>Number of bases that at least one read must overlap each uniquely aligned side of a predicted junction</td><td align="center">&ge;&nbsp;1</td></tr></table>
<h2>Read Alignment Evidence</h2>
<p>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th>option</th><th>value</th></tr><tr><td>Mode</td><td>Consensus/Mixed Base</td></tr><tr><td>Ploidy</td><td>1 (haploid)</td></tr><tr><td>Consensus mutation E-value cutoff</td><td>10</td></tr><tr><td>Consensus frequency cutoff</td><td>0.8</td></tr><tr><td>Consensus minimum variant coverage each strand</td><td>OFF</td></tr><tr><td>Consensus minimum total coverage each strand</td><td>OFF</td></tr><tr><td>Consensus minimum variant coverage</td><td>OFF</td></tr><tr><td>Consensus minimum total coverage</td><td>OFF</td></tr><tr><td>Polymorphism E-value cutoff</td><td>10</td></tr><tr><td>Polymorphism frequency cutoff</td><td>0.2</td></tr><tr><td>Polymorphism minimum variant coverage each strand</td><td>OFF</td></tr><tr><td>Polymorphism minimum total coverage each strand</td><td>OFF</td></tr><tr><td>Polymorphism minimum variant coverage</td><td>OFF</td></tr><tr><td>Polymorphism minimum total coverage</td><td>OFF</td></tr><tr><td>Polymorphism bias cutoff</td><td>OFF</td></tr><tr><td>Predict indel polymorphisms</td><td>YES</td></tr><tr><td>Skip indel polymorphisms in homopolymers runs of</td><td>OFF</td></tr><tr><td>Skip base substitutions when they create a homopolymer flanked on each side by</td><td>OFF</td></tr></table><p>
<h2>Software Versions</h2>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th>program</th><th>version</th></tr>
<tr><td>bowtie2</td><td>2.3.2</td></tr>
<tr><td>R</td><td>3.2.1</td></tr>
</table><p>
<h2>Execution Times</h2>
<table border="0" cellspacing="1" cellpadding="5">
<tr><th>step</th><th>start</th><th>end</th><th>elapsed</th></tr>
<tr><td>Read and reference sequence file input</td><td>00:52:34&nbsp;17&nbsp;May&nbsp;2019</td><td>00:53:50&nbsp;17&nbsp;May&nbsp;2019</td><td>1&nbsp;minute&nbsp;16&nbsp;seconds</td></tr>
<tr><td>Read alignment to reference genome</td><td>00:53:50&nbsp;17&nbsp;May&nbsp;2019</td><td>01:07:15&nbsp;17&nbsp;May&nbsp;2019</td><td>13&nbsp;minutes&nbsp;25&nbsp;seconds</td></tr>
<tr><td>Preprocessing alignments for candidate junction identification</td><td>01:07:15&nbsp;17&nbsp;May&nbsp;2019</td><td>01:07:45&nbsp;17&nbsp;May&nbsp;2019</td><td>30&nbsp;seconds</td></tr>
<tr><td>Preliminary analysis of coverage distribution</td><td>01:07:45&nbsp;17&nbsp;May&nbsp;2019</td><td>01:09:49&nbsp;17&nbsp;May&nbsp;2019</td><td>2&nbsp;minutes&nbsp;4&nbsp;seconds</td></tr>
<tr><td>Identifying junction candidates</td><td>01:09:49&nbsp;17&nbsp;May&nbsp;2019</td><td>01:09:50&nbsp;17&nbsp;May&nbsp;2019</td><td>1&nbsp;second</td></tr>
<tr><td>Re-alignment to junction candidates</td><td>01:09:50&nbsp;17&nbsp;May&nbsp;2019</td><td>01:14:05&nbsp;17&nbsp;May&nbsp;2019</td><td>4&nbsp;minutes&nbsp;15&nbsp;seconds</td></tr>
<tr><td>Resolving best read alignments</td><td>01:14:05&nbsp;17&nbsp;May&nbsp;2019</td><td>01:15:51&nbsp;17&nbsp;May&nbsp;2019</td><td>1&nbsp;minute&nbsp;46&nbsp;seconds</td></tr>
<tr><td>Creating BAM files</td><td>01:15:51&nbsp;17&nbsp;May&nbsp;2019</td><td>01:17:27&nbsp;17&nbsp;May&nbsp;2019</td><td>1&nbsp;minute&nbsp;36&nbsp;seconds</td></tr>
<tr><td>Tabulating error counts</td><td>01:17:27&nbsp;17&nbsp;May&nbsp;2019</td><td>01:17:52&nbsp;17&nbsp;May&nbsp;2019</td><td>25&nbsp;seconds</td></tr>
<tr><td>Re-calibrating base error rates</td><td>01:17:52&nbsp;17&nbsp;May&nbsp;2019</td><td>01:18:13&nbsp;17&nbsp;May&nbsp;2019</td><td>21&nbsp;seconds</td></tr>
<tr><td>Examining read alignment evidence</td><td>01:18:14&nbsp;17&nbsp;May&nbsp;2019</td><td>01:23:52&nbsp;17&nbsp;May&nbsp;2019</td><td>5&nbsp;minutes&nbsp;38&nbsp;seconds</td></tr>
<tr><td>Polymorphism statistics</td><td>01:23:52&nbsp;17&nbsp;May&nbsp;2019</td><td>01:23:53&nbsp;17&nbsp;May&nbsp;2019</td><td>1&nbsp;second</td></tr>
<tr><td>Output</td><td>01:23:53&nbsp;17&nbsp;May&nbsp;2019</td><td>01:24:41&nbsp;17&nbsp;May&nbsp;2019</td><td>48&nbsp;seconds</td></tr>
<tr class="highlight_table_row">
<td colspan="3" ><b>Total</b></td>
<td><b>32&nbsp;minutes&nbsp;6&nbsp;seconds</b></td>
</tr>
</table></body></html>